File: specifier.ts
export type Specifier = {
  proto?: string;
  scope?: string;
  pkg?: string;
  name?: string;
  path?: string;
};

const regexp = /^(?:@(\w[\w-.]*)\/)?(\w[\w-.]*)(\/(.*))?/;

/**
 * Parse a input string into a Specifier.
 *
 * input can be a npm package name with optional proto, subpath.
 */
export const parseSpecifier = (input: string): Specifier | null => {
  if (!input) {
    return null;
  }
  const sp = input.split(":");
  const path = sp.pop();

  const match = path.match(regexp);

  console.log(match);
  if (!match || !match[0] || !match[2]) {
    return null;
  }

  const scope = match[1];
  const pkg = match[2];
  const result: Specifier = {
    name: scope ? `@${scope}/${pkg}` : pkg,
    scope,
    pkg,
    proto: sp.pop(),
    path: match[4],
  };
  return result;
};

File: specifier.go
package resolve

import (
	"errors"
	"regexp"
)

type Specifier struct {
	Proto string
	Scope string
	Pkg   string
	Name  string
	Path  string
}

var (
	SpecifierRegex      = regexp.MustCompile(`^(?:([\w][\w0-9]*):)?(?:(@[a-z0-9-~][a-z0-9-._~]*)/)?([a-z0-9-][a-z0-9-._]*)(/([\w/-]*))?$`)
	ErrInvalidSpecifier = errors.New("resolve: invalid specifier")
)

func NewSpecifier(input string) (*Specifier, error) {
	if input == "" {
		return nil, ErrInvalidSpecifier
	}

	match := SpecifierRegex.FindStringSubmatch(input)
	if match == nil || match[0] == "" {
		return nil, ErrInvalidSpecifier
	}

	proto := match[1]
	scope := match[2]
	pkg := match[3]
	path := match[5]

	if pkg == "" {
		return nil, ErrInvalidSpecifier
	}

	var name string
	if scope != "" {
		name = scope + "/" + pkg
	} else {
		name = pkg
	}

	return &Specifier{
		Proto: proto,
		Scope: scope,
		Pkg:   pkg,
		Path:  path,
		Name:  name,
	}, nil
}
